### ISSUE TITLE
Application Startup Profiling

### ISSUE THREAD

**Issue by franz1981** [async-profiler/async-profiler#726]
While profiling an application startup by using async-profiling as an agent there's still the missing bit of info re measuring the runtime bootstrap till agent loading complete; I see this one as an interesting feature to be implemented together with the "classify" one i.e. marking as "bootstraps" the native-only stack-traces of the JVM before "jattach" is available. 
I have no idea if it's possible to implement something like that given the current architecture of the profiler, but it would be of great help for applications that need to be restarted/scale-up frequently, providing additional useful informations re hidden and still important costs (compiler threads startup cost, GC threads startup cost, initialization of GC data-structures and/or pre-touch costs, etc etc).

Obviously this means that it cannot be used as an agent anymore :)

**Comment by parttimenerd** [comment]
One idea would be to run `perf record` for the first 10 or so seconds on the JVM (till the async-profiler is surely loaded) and then combine the results later with the async-profiler results.

**Comment by parttimenerd** [comment]
Yes, but I think implementing it directly in the Agent is impossible. I rather envision something like

```sh
ap-record java -agentlib:libasyncprofiler.so=... 
```

Where ap-record is able to set environment variables, so that the agent knows where to look for the recording of the initial few seconds.

**Comment by apangin** [comment]
I have made a PoC recently of starting async-profiler as LD_PRELOAD'ed library (spoiler: for profiling native apps), so it's definitely possible to use it for JVM startup, too.
While not promising anything and not committing myself, I can tell this is one of the features that may appear in near future.

**Comment by parttimenerd** [comment]
I'm happy to help in the effort.

**Comment by franz1981** [comment]
And can schedule some free cycles to help as well, in case (although I know that @apangin is a fast train when he decides to do it :D )

**Comment by apangin** [comment]
Profiling of non-Java processes has been added in #751.
It can also be used for profiling `java` startup:

```
LD_PRELOAD=libasyncProfiler.so ASPROF_COMMAND=start,event=cpu,interval=100us,threads,file=profile.html java -version
```
![startup-native](https://github.com/async-profiler/async-profiler/assets/1749416/9b4e5e65-45c7-4911-a296-1d9db9f5b36d)

The downside of such a profile is that it does not name any Java frames.

Java frame information becomes available after JVM bootstrap has completed, i.e. after JVM TI [VMStart](https://docs.oracle.com/en/java/javase/20/docs/specs/jvmti.html#VMStart) even has been sent.
A recently added `cstack=vm` mode is also helpful for analyzing JVM startup from that moment.

![startup-java](https://github.com/async-profiler/async-profiler/assets/1749416/99fcf05b-f42d-48d0-b3bf-2ee3ffc6739e)

Merging two above profiles is possible, but adds redundant complexity/ugliness to the code. I don't think it's worth productizing the feature right now, assuming the existing capabilities are enough to troubleshoot startup issues.

**Comment by krk** [comment]
With LD_PRELOAD way, async-profiler can profile an app before JVM is started.

Closing this issue. Feel free to reopen or file a new issue, if it doesn't meet the use case here.

**Comment by apangin** [comment]
This issue still makes sense, as there are two opportunities to make startup profiling better:
1. With `LD_PRELOAD`, async-profiler does not currently make use of JVM-specific capabilities: it does not show Java method frames and cannot profile heap allocations or contended locks. It would be nice if in `LD_PRELOAD` mode it could automatically detect JVM start and switch to agent mode. However, a tradeoff between complexity of the solution and its impact makes me think it's not really worth doing.
2. But another thing that I believe *is* worth implementing is a support for [can_generate_early_vmstart](https://docs.oracle.com/en/java/javase/21/docs/specs/jvmti.html#jvmtiCapabilities.can_generate_early_vmstart) capability. Without it, profiler starts after the module system is fully initialized with can be rather late wrt. application startup. See [VMStart](https://docs.oracle.com/en/java/javase/21/docs/specs/jvmti.html#VMStart) documentation for details.

**Comment by fandreuz** [comment]
I did some thinking on the second point. There are two problems with moving profiler initialization in `VMStart`:
- Generation of jmethodIDs: we can't use [GetLoadedClasses](https://docs.oracle.com/en/java/javase/21/docs/specs/jvmti.html#GetLoadedClasses) in `VMStart`, and I couldn't find a way to retrieve all loaded classes in `java.base`. I thought about enabling `ClassLoad` events, but of course that wouldn't give us what was loaded earlier. We could hardcode a list of classes, and then use `FindClass` on each of them in the `VMStart` callback: if we get a valid `jclass` we pass it to `loadMethodIDs`. We then keep the call to `loadAllMethodIDs` in `VMInit` to catch the remaining classes.
- `initThreadBridge`: [GetCurrentThread](https://docs.oracle.com/en/java/javase/21/docs/specs/jvmti.html#GetCurrentThread) can be called before live phase, but [not in early start phase](https://bugs.openjdk.org/browse/JDK-8172969). I believe this is not a strong blocker, but some code paths (e.g. AsyncGetCallTrace) might not work before the thread bridge is initialized.

@apangin what do you think about the first point?

**Comment by apangin** [comment]
> I thought about enabling ClassLoad events, but of course that wouldn't give us what was loaded earlier.

JVMTI agents are loaded early at JVM startup when no Java classes have been loaded yet.

> some code paths (e.g. AsyncGetCallTrace) might not work before the thread bridge is initialized

It's OK if AsyncGetCallTrace does not work during JVMTI_PHASE_START; only `walkVM` is relevant for this feature.

**Comment by fandreuz** [comment]
@apangin I've got a draft branch for moving profiler start to `VMStart`. I've got [some problems](https://github.com/fandreuz/async-profiler/actions/runs/20977953096) on JDK8 though. The segfaults are caused by JNI calls (`FindClass`, as part of `TSC::enable`) in `VMStart`, which is [not safe](https://github.com/openjdk/jdk/blob/45990d796ffafc228c6e843049c80aefedb0f12b/src/hotspot/share/prims/jvmti.xml#L13474). It seems to be working fine in JDK8+, but I'm wondering if we should rely on this undocumented behavior?

We can't postpone `TSC::enable` since we need `TSC::ticks` to record a JFR event. Perhaps we could record the events anyway, and amend the time after `VMInit`?
